src/rpc/rpc.cc

```diff
@@ -7,25 +7,258 @@
+    network::Message ndarray_meta_msg;
+    ndarray_meta_msg.data = ndarray_meta_buffer;
+    ndarray_meta_msg.size = zerocopy_blob.size();
+    ndarray_meta_msg.deallocator = network::DefaultMessageDeleter;
+    CHECK_EQ(RPCContext::ThreadLocal()->sender->Send(
+      ndarray_meta_msg, msg.server_id), ADD_SUCCESS);
+    // send ndarray count
+    char* ndarray_count = new char[sizeof(int32_t)];
+    *(reinterpret_cast<int32_t*>(ndarray_count)) = msg.tensors.size();
+    network::Message ndarray_count_msg;
+    ndarray_count_msg.data = ndarray_count;
+    ndarray_count_msg.size = sizeof(int32_t);
+    ndarray_count_msg.deallocator = network::DefaultMessageDeleter;
+    CHECK_EQ(RPCContext::ThreadLocal()->sender->Send(
+      ndarray_count_msg, msg.server_id), ADD_SUCCESS);
+    // send real ndarray data
+    for (auto ptr : zc_write_strm.buffer_list()) {
+      network::Message ndarray_data_msg;
+      ndarray_data_msg.data = reinterpret_cast<char*>(ptr.data);
+      ndarray_data_msg.size = ptr.size;
+      NDArray tensor = ptr.tensor;
+      ndarray_data_msg.deallocator = [tensor](network::Message*) {};
+      CHECK_EQ(RPCContext::ThreadLocal()->sender->Send(
+        ndarray_data_msg, msg.server_id), ADD_SUCCESS);
+    }
+  }
   return kRPCSuccess;
 }
 
 RPCStatus RecvRPCMessage(RPCMessage* msg, int32_t timeout) {
-  // TODO(???): Implement
+  // ignore timeout now
+  network::Message rpc_meta_msg;
+  int send_id;
+  CHECK_EQ(RPCContext::ThreadLocal()->receiver->Recv(
+    &rpc_meta_msg, &send_id), REMOVE_SUCCESS);
+  bool has_tensor = DeserializeRPCMeta(msg, rpc_meta_msg.data, rpc_meta_msg.size);
+  if (has_tensor) {
+    // Recv ndarray meta data
+    network::Message ndarray_meta_msg;
+    CHECK_EQ(RPCContext::ThreadLocal()->receiver->RecvFrom(
+      &ndarray_meta_msg, send_id), REMOVE_SUCCESS);
+    std::string zerocopy_blob;
+    zerocopy_blob.resize(ndarray_meta_msg.size);
+    memcpy(const_cast<char*>(zerocopy_blob.data()),
+           ndarray_meta_msg.data,
+           ndarray_meta_msg.size);
+    ndarray_meta_msg.deallocator(&ndarray_meta_msg);
+    // Recv ndarray count
+    network::Message ndarray_count_msg;
+    CHECK_EQ(RPCContext::ThreadLocal()->receiver->RecvFrom(
+      &ndarray_count_msg, send_id), REMOVE_SUCCESS);
+    int32_t ndarray_count = *(reinterpret_cast<int32_t*>(ndarray_count_msg.data));
+    ndarray_count_msg.deallocator(&ndarray_count_msg);
+    // Recv real ndarray data
+    std::vector<void* > buffer_list(ndarray_count);
+    for (int i = 0; i < ndarray_count; ++i) {
+      network::Message ndarray_data_msg;
+      CHECK_EQ(RPCContext::ThreadLocal()->receiver->RecvFrom(
+        &ndarray_data_msg, send_id), REMOVE_SUCCESS);
+      buffer_list[i] = ndarray_data_msg.data;
+    }
+    StringStreamWithBuffer zc_read_strm(&zerocopy_blob, buffer_list);
+    msg->tensors.resize(ndarray_count);
+    static_cast<dmlc::Stream *>(&zc_read_strm)->Read(&msg->tensors);
```

Comment:
who owns the memory of the tensors?

Reply:
The `msg->tensors` will own all the memory. I think this code is done by JJ's `Read(&msg->tensors);`